File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-15.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();
        Deque<TreeNode> queue = new ArrayDeque<>();  
        queue.addFirst(root);

        while(!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> levelNodes = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.removeFirst();
                levelNodes.add(node.val);
                if (node.left != null) queue.addLast(node.left);
                if (node.right != null) queue.addLast(node.right);
            }

            res.add(levelNodes);

        }
        
        return res;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-15.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.isValid(root, float('-inf'), float('inf'))

    def isValid(self, current: Optional[TreeNode], left: float, right: float) -> bool:

        if not current: return True

        elif current.val <= left or current.val >= right: return False

        # on the left side the upper bound will change
        # on the right side the lower bound will change
        return self.isValid(current.left, left, current.val) and \
               self.isValid(current.right, current.val, right) 
